File: Day_11/Day11.py
from copy import deepcopy


def flash(board, row, col, flashboard):
    if row < 0 or row >= 10 or col < 0 or col >= 10 or flashboard[row][col]:
        return

    board[row][col] += 1
    if board[row][col] > 9:
        flashboard[row][col] = True
        board[row][col] = 0
        # Bottom
        flash(board, row + 1, col, flashboard)
        # Top
        flash(board, row - 1, col, flashboard)
        # Right
        flash(board, row, col + 1, flashboard)
        # Left
        flash(board, row, col - 1, flashboard)
        # Bottom - Right
        flash(board, row + 1, col + 1, flashboard)
        # Bottom - Left
        flash(board, row + 1, col - 1, flashboard)
        # Top - Left
        flash(board, row - 1, col + 1, flashboard)
        # Top - Left
        flash(board, row - 1, col - 1, flashboard)


def simulateFlashing(board):
    flashCount = 0
    for step in range(1, 101):
        flashboard = [[False for _ in range(0, 10)] for _ in range(0, 10)]
        for row in range(0, 10):
            for col in range(0, 10):
                flash(board, row, col, flashboard)

        currFlashCount = 0
        for row in range(0, 10):
            for col in range(0, 10):
                if board[row][col] == 0:
                    currFlashCount += 1
        flashCount += currFlashCount
    return flashCount


def simultaneousFlashStep(board):
    step = 0
    while True:
        step += 1
        flashboard = [[False for _ in range(0, 10)] for _ in range(0, 10)]
        for row in range(0, 10):
            for col in range(0, 10):
                flash(board, row, col, flashboard)

        currFlashCount = 0
        for row in range(0, 10):
            for col in range(0, 10):
                if board[row][col] == 0:
                    currFlashCount += 1

        if currFlashCount == 100:
            return step


def getInput():
    board = []
    with open("./Day11_input.txt") as inputFile:
        for line in inputFile:
            if '\n' in line:
                line = line.strip()
            board.append([int(x) for x in line])
    return board


board = getInput()
copyBoard = deepcopy(board)

# Part 1
print("Number of Flashes in 100 Steps: " + str(simulateFlashing(board)))

# Part 2
print("First Synchronous Flash occurs at: " +
      str(simultaneousFlashStep(copyBoard)))

File: Day_11/Day11.cpp
#include <iostream>
#include <fstream>
#include <vector>
using namespace std;

void flash(vector<vector<int>> &board, int row, int col, vector<vector<bool>> &flashboard, long &flashCount)
{
    if (row < 0 || row >= 10 || col < 0 || col >= 10 || flashboard[row][col])
        return;

    board[row][col] += 1;
    if (board[row][col] > 9)
    {
        flashboard[row][col] = true;
        board[row][col] = 0;
        flashCount += 1;
        // Bottom
        flash(board, row + 1, col, flashboard, flashCount);
        // Top
        flash(board, row - 1, col, flashboard, flashCount);
        // Right
        flash(board, row, col + 1, flashboard, flashCount);
        // Left
        flash(board, row, col - 1, flashboard, flashCount);
        // Bottom - Right
        flash(board, row + 1, col + 1, flashboard, flashCount);
        // Bottom - Left
        flash(board, row + 1, col - 1, flashboard, flashCount);
        // Top - Left
        flash(board, row - 1, col + 1, flashboard, flashCount);
        // Top - Left
        flash(board, row - 1, col - 1, flashboard, flashCount);
    }
}

long simulateFlashing(vector<vector<int>> board)
{
    long flashCount = 0;
    for (int step = 1; step <= 100; step++)
    {
        long currFlashCount = 0;
        vector<vector<bool>> flashboard(10, vector<bool>(10, false));
        for (int row = 0; row < 10; row++)
            for (int col = 0; col < 10; col++)
                flash(board, row, col, flashboard, currFlashCount);

        flashCount += currFlashCount;
    }
    return flashCount;
}

int simultaneousFlashStep(vector<vector<int>> board)
{
    int step = 0;
    while (true)
    {
        step++;
        long currFlashCount = 0;
        vector<vector<bool>> flashboard(10, vector<bool>(10, false));
        for (int row = 0; row < 10; row++)
            for (int col = 0; col < 10; col++)
                flash(board, row, col, flashboard, currFlashCount);

        if (currFlashCount == 100)
            return step;
    }
    return 0;
}

vector<vector<int>> getInput()
{
    ifstream inputFile("./Day11_Input.txt");
    string line;
    vector<vector<int>> board;
    while (getline(inputFile, line))
    {
        vector<int> row;
        for (char ch : line)
            row.push_back(ch - '0');

        board.push_back(row);
    }
    return board;
}

int main()
{
    vector<vector<int>> board = getInput();

    // Part 1
    cout << "Number of Flashes in 100 Steps: " << simulateFlashing(board) << endl;

    // Part 2
    cout << "First Synchronous Flash occurs at: " << simultaneousFlashStep(board) << endl;

    return 0;
}